### ISSUE TITLE
New dabbrev completions too eager

### ISSUE THREAD

**Issue by geraldus** [haskell/haskell-mode#741]
Currently I'm writing some GHCJS code and didn't setup interactive version yet, so I always have dabbrev completions. With 10-20 opened buffers dabbrev completion method becomes too eager already, i.e. completions takes too long. It is possible to limit dabbrev list of buffers to search completions. For example, try to lookup nearest folder contains .git, or .cabal file, set some haskell-completions-top-folder local var, and when searching dynamic abbrevs take this var in account. When session is available, search for d-abbrevs in buffers assigned to current session.

What do you think?

**Comment by gracjan** [comment]
I use dabbrev by M-/ command key. I would love to have M-/ do dabbrev,
completion kept only for more context-sensitive semantic information only.

**Comment by geraldus** [comment]
If you don't use **company-mode** I think you does not encounter this issue.  Here is a little [demo](https://drive.google.com/open?id=0BwLfIDb7xEV7d3Vzd1FkNWh6Z1k).  Company mode, no REPL (hence completions are always comes from dynamic abbrevs), 50 buffers opened.

**Comment by geraldus** [comment]
If someone want to fix this before I find a time this might be helpful [dabbrev.el](https://github.com/emacs-mirror/emacs/blob/master/lisp/dabbrev.el).

**Comment by gracjan** [comment]
@geraldus: Why do you use `dabbrev--reset-global-variables`? It is here:

https://github.com/haskell/haskell-mode/blob/a8146714e12b2af6afe1a086636dcc4ee6bb7825/haskell-completions.el#L272

As I understand `dabbrev` has some caching mechanism to speed up expansions, resetting kills this optimization. Or am I mistaken?

**Comment by gracjan** [comment]
@geraldus: My point about M-/ was that `company-mode` should NOT use dabbrev as source. I envision it like this:

- `company-mode` uses only contextual sources for completions: `:complete`, `PRAGMA`, `import`
- M-/ uses non-contextual sources for completions: abbrev, dabbrev

So 'smart' stuff is for company/completion, non-smart for dabbrev. Am I mistaken?

**Comment by cocreature** [comment]
@gracjan I agree if people want to use dabbrev via company mode they can still add it themselves but I am not sure what we gain by doing that from haskell mode.

**Comment by geraldus** [comment]
Sometimes REPL is not running for some reason (in my case there is no GHCJSi available yet) or user just does not want to run it, that is, `:complete` command is not available too.  So how company should look up completions?

**Comment by gracjan** [comment]
@geraldus: My (original) company setup has this part:

```
...
Company Backends:
INS DEL Back-end: Value Menu completion-at-point-functions
INS DEL Back-end: Value Menu Merged Back-ends:
            INS DEL Back-end: Value Menu dabbrev for code
            INS DEL Back-end: Value Menu GNU Global
            INS DEL Back-end: Value Menu etags
            INS DEL Back-end: Value Menu Programming language keywords
            INS
INS DEL Back-end: Value Menu Oddmuse
INS DEL Back-end: Value Menu Files
INS DEL Back-end: Value Menu dabbrev for plain text
```

Does this mean `dabbrev` is just handled by `company-mode` and there is no need to do anything about it inside `haskell-mode`?

**Comment by gracjan** [comment]
@geraldus: If interpreter is not running then `:complete` is not an available source. What is left is of the semantic sources, currently PRAGMA, LANGUAGE and that is enough. Someday we might add more sources

`dabbrev` is available always as M-/, so nothing is lost.

**Comment by gracjan** [comment]
@cocreature: How do they 'add it themselves'? It is of vital importance to me that stuff works out of the box as it should.

**Comment by cocreature** [comment]
@gracjan IIRC there is a company-dabbrev backend built into company mode so you can just add that to your backend. Looking at it  we don't seem to provide a company backend atm, so I am not quite sure how to easily combine them. Of course it is important that stuff works out of the box, but to be frank the current dabbrev integration doesn't work (well enough). Mostly it hangs up my emacs.

**Comment by gracjan** [comment]
In that case we should make sure that `company-dabbrev` works with
`haskell-mode` in case somebody decides to use it.

`haskell-mode` should NOT do anything special with dabbrev because this is
code duplication and is problematic to get right.

**Comment by geraldus** [comment]
Well, I think we should have some overrides for company's dabbrev source, for example case sensitivity.  And if it is possible we should have some notes about completions itself and `company-mode` particularly.

**Comment by gracjan** [comment]
We have some configuration for dabbrev in haskell-mode already.

**Comment by geraldus** [comment]
Ok, to test this I need to rewrite some code in completion function itself, I'll update pull request soon.

**Comment by cocreature** [comment]
Could we just disable dabbrev until this is fixed? It hangs up my emacs for multiple seconds which is for obvious reasons pretty annoying. I'll be happy to send a pullrequest for that if you agree.

**Comment by geraldus** [comment]
Please excuse me my delay and inconveniences brought by last changes.  I have some troubles here and can do nothing at the moment.  I will continue as soon as possible!  Go ahead!

**Comment by geraldus** [comment]
@cocreature haskell-mode have some configuration for `dabbrev` already, by the way.  When you'll disable my stuff, can you please test how it works and share your feelings?  The one thing I've already mentioned is that dabbrev ignore case making completions useless in most cases.  In the end, I think it is possible to remove dabbrev-stuff which I provided replacing it by better configuration to make everybody happy out-of-box (when REPL completions are no available)/

**Comment by gracjan** [comment]
Pull request disabling dabbrev welcome.

**Comment by gracjan** [comment]
Fixed by #838.
